src/node.rs

```diff
@@ -114,16 +126,263 @@ impl InMemoryNodeInner {
+        &self,
+        req: zksync_types::transaction_request::CallRequest,
+    ) -> jsonrpc_core::Result<Fee> {
+        let mut l2_tx = match l2_tx_from_call_req(req, MAX_TX_SIZE) {
+            Ok(tx) => tx,
+            Err(e) => {
+                let error = Web3Error::SerializationError(e);
+                return Err(into_jsrpc_error(error));
+            }
+        };
+
+        let tx: Transaction = l2_tx.clone().into();
+        let fair_l2_gas_price = L2_GAS_PRICE;
+
+        // Calculate Adjusted L1 Price
+        let gas_price_scale_factor = 1.2;
+        let l1_gas_price = {
+            let current_l1_gas_price = ((self.l1_gas_price as f64) * gas_price_scale_factor) as u64;
+
+            // In order for execution to pass smoothly, we need to ensure that block's required gasPerPubdata will be
+            // <= to the one in the transaction itself.
+            adjust_l1_gas_price_for_tx(
+                current_l1_gas_price,
+                L2_GAS_PRICE,
+                tx.gas_per_pubdata_byte_limit(),
+            )
+        };
+
+        let (base_fee, gas_per_pubdata_byte) =
+            derive_base_fee_and_gas_per_pubdata(l1_gas_price, fair_l2_gas_price);
+
+        // Check for properly formatted signature
+        if l2_tx.common_data.signature.is_empty() {
+            l2_tx.common_data.signature = vec![0u8; 65];
+            l2_tx.common_data.signature[64] = 27;
+        }
+
+        l2_tx.common_data.fee.gas_per_pubdata_limit = MAX_GAS_PER_PUBDATA_BYTE.into();
+        l2_tx.common_data.fee.max_fee_per_gas = base_fee.into();
+        l2_tx.common_data.fee.max_priority_fee_per_gas = base_fee.into();
+
+        let mut storage_view = StorageView::new(&self.fork_storage);
+
+        // Calculate gas_for_bytecodes_pubdata
+        let pubdata_for_factory_deps = l2_tx
+            .execute
+            .factory_deps
+            .as_deref()
+            .unwrap_or_default()
+            .iter()
+            .map(|bytecode| {
+                if storage_view.is_bytecode_known(&hash_bytecode(bytecode)) {
+                    return 0;
+                }
+
+                let length = if let Ok(compressed) = compress_bytecode(bytecode) {
+                    compressed.len()
+                } else {
+                    bytecode.len()
+                };
+                let publish_byte_overhead = 100;
+                length as u32 + publish_byte_overhead
+            })
+            .sum::<u32>();
+
+        if pubdata_for_factory_deps > MAX_PUBDATA_PER_BLOCK {
+            return Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                "exceeds limit for published pubdata".into(),
+                Default::default(),
+            )));
+        }
+
+        let gas_for_bytecodes_pubdata: u32 =
+            pubdata_for_factory_deps * (gas_per_pubdata_byte as u32);
+
+        // We are using binary search to find the minimal values of gas_limit under which the transaction succeeds
```

Comment:
Might be worth splitting the binary search into own fn. The `estimate_gas_impl` is quite long so I am just thinking of where we can split out isolated functionality. 🤷

Reply:
I tried that and just ran into more problems. The struggle is this structure is how it's implemented in `zksync-era`, so was trying to keep it close to the source.

Will give it another shot now that I have a more "finalized" version working

Reply:
After trying some refactoring, the binary search is actually a relatively small part of this function. I'm thinking to leave it as is (for now), as I have a feeling we'll be comparing this code to `zksync-era` sooner than later :(